### PR TITLE
CWMS-866 Added re-try of DirectX509 token if refresh token is expired.

### PR DESCRIPTION
**Pull request by rma-bryson**

### CODE REVIEW

File: cwms-http-client/src/main/java/mil/army/usace/hec/cwms/http/client/auth/OAuth2TokenProvider.java [hunk 1]
```diff
@@ -8,4 +8,6 @@ public interface OAuth2TokenProvider {
 
     OAuth2Token refreshToken() throws IOException;
 
+    OAuth2Token getDirectX509Token() throws IOException;
```
Comment (adamkorynta) [inline cwms-http-client/src/main/java/mil/army/usace/hec/cwms/http/client/auth/OAuth2TokenProvider.java:11]: Do we care if this is X509? Also, do we expect this to be a new instance? If so, I'd think renaming to `newToken()` would be more appropriate. That would distinguish from `refreshToken()`.

File: cwms-http-client/src/main/java/mil/army/usace/hec/cwms/http/client/auth/AccessTokenValidator.java [hunk 2]
```diff
@@ -0,0 +1,37 @@
+package mil.army.usace.hec.cwms.http.client.auth;
+
+import com.auth0.jwt.JWT;
+import com.auth0.jwt.interfaces.DecodedJWT;
+import java.time.Instant;
+import java.util.Date;
+
+public final class AccessTokenValidator {
```
Comment (adamkorynta) [inline cwms-http-client/src/main/java/mil/army/usace/hec/cwms/http/client/auth/AccessTokenValidator.java:8]: What is the use-case outside this library?

File: cwms-http-client/src/main/java/mil/army/usace/hec/cwms/http/client/auth/AccessTokenValidator.java [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+
+    private static final String REFRESH_EXPIRED_BUFFER_PROPERTY_KEY = "cwms.http.client.token.refresh.buffer.millis";
```
Comment (adamkorynta) [inline cwms-http-client/src/main/java/mil/army/usace/hec/cwms/http/client/auth/AccessTokenValidator.java:10]: I don't think any users would set milis less than a second.

File: cwms-http-client/src/main/java/mil/army/usace/hec/cwms/http/client/auth/AccessTokenValidator.java [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+    private static final long DEFAULT_REFRESH_EXPIRED_BUFFER_SECONDS = 1;
```
Comment (adamkorynta) [inline cwms-http-client/src/main/java/mil/army/usace/hec/cwms/http/client/auth/AccessTokenValidator.java:11]: I would bump this up to a minute for safety. The refresh token expiration for cumulus is on the order of hours right?

File: cwms-http-client/src/main/java/mil/army/usace/hec/cwms/http/client/OAuth2TokenAuthenticator.java [hunk 5]
```diff
@@ -50,4 +55,15 @@ Request newRequestWithAccessTokenAsHeader(Response response, OAuth2Token oauth2T
             .addHeader(AUTHORIZATION_HEADER, oauth2Token.getTokenType() + " " + oauth2Token.getAccessToken())
             .build();
     }
+
+    private void validateX509Token(OAuth2Token updatedToken) throws IOException {
```
Comment (adamkorynta) [inline cwms-http-client/src/main/java/mil/army/usace/hec/cwms/http/client/OAuth2TokenAuthenticator.java:59]: I don't think we care about X509 here

**Review by adamkorynta — CHANGES_REQUESTED** [review]

File: cwms-http-client/src/main/java/mil/army/usace/hec/cwms/http/client/auth/OAuth2TokenProvider.java (hunk 1 above)
Comment (rma-bryson) [reply cwms-http-client/src/main/java/mil/army/usace/hec/cwms/http/client/auth/OAuth2TokenProvider.java:11]: No we don't care. I updated to newToken()

File: cwms-http-client/src/main/java/mil/army/usace/hec/cwms/http/client/OAuth2TokenAuthenticator.java (hunk 5 above)
Comment (rma-bryson) [reply cwms-http-client/src/main/java/mil/army/usace/hec/cwms/http/client/OAuth2TokenAuthenticator.java:58]: updated to newToken

**Review by adamkorynta — APPROVED** [review]

**Pull request merged**